README.md
# base58check

Base58Check and regular Base58 encoding implementation for Dart using dart:convert interface.


CHANGELOG.md
# Changelog

## 2.0.0
- Migrate to null safety.

## 1.0.1

Support Dart v2.

## 1.0.0

- Simplified interface to remove external SHA-256 requirement
- Created Base58CheckCodec.bitcoin() constructor
- Made Base58CheckCodec store instances of the encoder and decoder
  for performance improvement

## 0.1.1

- Updated to new collection and crypto interfaces

## 0.1.0

- Initial version


test/test_utils.dart
import "dart:typed_data";

const String _BYTE_ALPHABET = "0123456789abcdef";
List<int> hexToBytes(String hex) {
  hex = hex.replaceAll(" ", "");
  hex = hex.toLowerCase();
  if (hex.length % 2 != 0) hex = "0" + hex;
  Uint8List result = new Uint8List(hex.length ~/ 2);
  for (int i = 0; i < result.length; i++) {
    int value = (_BYTE_ALPHABET.indexOf(hex[i * 2]) << 4) //= byte[0] * 16
        +
        _BYTE_ALPHABET.indexOf(hex[i * 2 + 1]);
    result[i] = value;
  }
  return result;
}


test/base58_test.dart
library base58check.test.base58;

import "package:test/test.dart";

import "package:base58check/base58.dart";

import "test_utils.dart";

const String _bitcoinAlphabet =
    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
List<List<String>> _bitcoinVectors = [
  ["", ""],
  ["61", "2g"],
  ["626262", "a3gV"],
  ["636363", "aPEr"],
  ["73696d706c792061206c6f6e6720737472696e67", "2cFupjhnEsSn59qHXstmK2ffpLv2"],
  [
    "00eb15231dfceb60925886b67d065299925915aeb172c06647",
    "1NS17iag9jJgTHD1VXjvLCEnZuQ3rJDE9L"
  ],
  ["516b6fcd0f", "ABnLTmg"],
  ["bf4f89001e670274dd", "3SEo3LWLoPntC"],
  ["572e4794", "3EFU7m"],
  ["ecac89cad93923c02321", "EJDM8drfXA6uyA"],
  ["10c8511e", "Rt5zm"],
  ["00000000000000000000", "1111111111"]
];

void main() {
  group("base58check.base58", () {
    test("bitcoin_vectors", () {
      Base58Codec codec = const Base58Codec(_bitcoinAlphabet);
      for (List<String> vector in _bitcoinVectors) {
        List<int> bytes = hexToBytes(vector[0]);
        String encoding = vector[1];
        expect(codec.encode(bytes), equals(encoding),
            reason: "encode " + vector.toString());
        expect(codec.decode(encoding), equals(bytes),
            reason: "decode " + vector.toString());
      }
    });
  });
}


test/base58check_test.dart
library base58check.test.base58check;

import "package:test/test.dart";

import "package:base58check/base58check.dart";

import "test_utils.dart";

// vectors of correct bitcoin strings
List<List<String>> _bitcoinVectors = [
  [
    "1AGNa15ZQXAZUgFiqJ2i7Z2DPU2J6hW62i",
    "65a16059864a2fdbc7c99a4723a8395bc6f188eb",
    "0"
  ],
  [
    "3CMNFxN1oHBc4R1EpboAL5yzHGgE611Xou",
    "74f209f6ea907e2ea48f74fae05782ae8a665257",
    "5"
  ],
  [
    "mo9ncXisMeAoXwqcV5EWuyncbmCcQN4rVs",
    "53c0307d6851aa0ce7825ba883c6bd9ad242b486",
    "111"
  ],
  [
    "2N2JD6wb56AfK4tfmM6PwdVmoYk2dCKf4Br",
    "6349a418fc4578d10a372b54b45c280cc8c4382f",
    "196"
  ],
  [
    "5Kd3NBUAdUnhyzenEwVLy9pBKxSwXvE9FMPyR4UKZvpe6E3AgLr",
    "eddbdc1168f1daeadbd3e44c1e3f8f5a284c2029f78ad26af98583a499de5b19",
    "128"
  ],
  [
    "9213qJab2HNEpMpYNBa7wHGFKKbkDn24jpANDs2huN3yi4J11ko",
    "36cb93b9ab1bdabf7fb9f2c04f1b9cc879933530ae7842398eef5a63a56800c2",
    "239"
  ],
  [
    "1Ax4gZtb7gAit2TivwejZHYtNNLT18PUXJ",
    "6d23156cbbdcc82a5a47eee4c2c7c583c18b6bf4",
    "0"
  ],
  [
    "3QjYXhTkvuj8qPaXHTTWb5wjXhdsLAAWVy",
    "fcc5460dd6e2487c7d75b1963625da0e8f4c5975",
    "5"
  ],
  [
    "n3ZddxzLvAY9o7184TB4c6FJasAybsw4HZ",
    "f1d470f9b02370fdec2e6b708b08ac431bf7a5f7",
    "111"
  ],
  [
    "2NBFNJTktNa7GZusGbDbGKRZTxdK9VVez3n",
    "c579342c2c4c9220205e2cdc285617040c924a0a",
    "196"
  ],
  [
    "5K494XZwps2bGyeL71pWid4noiSNA2cfCibrvRWqcHSptoFn7rc",
    "a326b95ebae30164217d7a7f57d72ab2b54e3be64928a19da0210b9568d4015e",
    "128"
  ],
  [
    "93DVKyFYwSN6wEo3E2fCrFPUp17FtrtNi2Lf7n4G3garFb16CRj",
    "d6bca256b5abc5602ec2e1c121a08b0da2556587430bcf7e1898af2224885203",
    "239"
  ],
  [
    "1C5bSj1iEGUgSTbziymG7Cn18ENQuT36vv",
    "7987ccaa53d02c8873487ef919677cd3db7a6912",
    "0"
  ],
  [
    "3AnNxabYGoTxYiTEZwFEnerUoeFXK2Zoks",
    "63bcc565f9e68ee0189dd5cc67f1b0e5f02f45cb",
    "5"
  ],
  [
    "n3LnJXCqbPjghuVs8ph9CYsAe4Sh4j97wk",
    "ef66444b5b17f14e8fae6e7e19b045a78c54fd79",
    "111"
  ],
  [
    "2NB72XtkjpnATMggui83aEtPawyyKvnbX2o",
    "c3e55fceceaa4391ed2a9677f4a4d34eacd021a0",
    "196"
  ],
  [
    "5KaBW9vNtWNhc3ZEDyNCiXLPdVPHCikRxSBWwV9NrpLLa4LsXi9",
    "e75d936d56377f432f404aabb406601f892fd49da90eb6ac558a733c93b47252",
    "128"
  ],
  [
    "927CnUkUbasYtDwYwVn2j8GdTuACNnKkjZ1rpZd2yBB1CLcnXpo",
    "44c4f6a096eac5238291a94cc24c01e3b19b8d8cef72874a079e00a242237a52",
    "239"
  ],
  [
    "1Gqk4Tv79P91Cc1STQtU3s1W6277M2CVWu",
    "adc1cc2081a27206fae25792f28bbc55b831549d",
    "0"
  ],
  [
    "33vt8ViH5jsr115AGkW6cEmEz9MpvJSwDk",
    "188f91a931947eddd7432d6e614387e32b244709",
    "5"
  ],
  [
    "mhaMcBxNh5cqXm4aTQ6EcVbKtfL6LGyK2H",
    "1694f5bc1a7295b600f40018a618a6ea48eeb498",
    "111"
  ],
  [
    "2MxgPqX1iThW3oZVk9KoFcE5M4JpiETssVN",
    "3b9b3fd7a50d4f08d1a5b0f62f644fa7115ae2f3",
    "196"
  ],
  [
    "5HtH6GdcwCJA4ggWEL1B3jzBBUB8HPiBi9SBc5h9i4Wk4PSeApR",
    "091035445ef105fa1bb125eccfb1882f3fe69592265956ade751fd095033d8d0",
    "128"
  ],
  [
    "92xFEve1Z9N8Z641KQQS7ByCSb8kGjsDzw6fAmjHN1LZGKQXyMq",
    "b4204389cef18bbe2b353623cbf93e8678fbc92a475b664ae98ed594e6cf0856",
    "239"
  ],
  [
    "1JwMWBVLtiqtscbaRHai4pqHokhFCbtoB4",
    "c4c1b72491ede1eedaca00618407ee0b772cad0d",
    "0"
  ],
  [
    "3QCzvfL4ZRvmJFiWWBVwxfdaNBT8EtxB5y",
    "f6fe69bcb548a829cce4c57bf6fff8af3a5981f9",
    "5"
  ],
  [
    "mizXiucXRCsEriQCHUkCqef9ph9qtPbZZ6",
    "261f83568a098a8638844bd7aeca039d5f2352c0",
    "111"
  ],
  [
    "2NEWDzHWwY5ZZp8CQWbB7ouNMLqCia6YRda",
    "e930e1834a4d234702773951d627cce82fbb5d2e",
    "196"
  ],
  [
    "5KQmDryMNDcisTzRp3zEq9e4awRmJrEVU1j5vFRTKpRNYPqYrMg",
    "d1fab7ab7385ad26872237f1eb9789aa25cc986bacc695e07ac571d6cdac8bc0",
    "128"
  ],
  [
    "91cTVUcgydqyZLgaANpf1fvL55FH53QMm4BsnCADVNYuWuqdVys",
    "037f4192c630f399d9271e26c575269b1d15be553ea1a7217f0cb8513cef41cb",
    "239"
  ],
  [
    "19dcawoKcZdQz365WpXWMhX6QCUpR9SY4r",
    "5eadaf9bb7121f0f192561a5a62f5e5f54210292",
    "0"
  ],
  [
    "37Sp6Rv3y4kVd1nQ1JV5pfqXccHNyZm1x3",
    "3f210e7277c899c3a155cc1c90f4106cbddeec6e",
    "5"
  ],
  [
    "myoqcgYiehufrsnnkqdqbp69dddVDMopJu",
    "c8a3c2a09a298592c3e180f02487cd91ba3400b5",
    "111"
  ],
  [
    "2N7FuwuUuoTBrDFdrAZ9KxBmtqMLxce9i1C",
    "99b31df7c9068d1481b596578ddbb4d3bd90baeb",
    "196"
  ],
  [
    "5KL6zEaMtPRXZKo1bbMq7JDjjo1bJuQcsgL33je3oY8uSJCR5b4",
    "c7666842503db6dc6ea061f092cfb9c388448629a6fe868d068c42a488b478ae",
    "128"
  ],
  [
    "93N87D6uxSBzwXvpokpzg8FFmfQPmvX4xHoWQe3pLdYpbiwT5YV",
    "ea577acfb5d1d14d3b7b195c321566f12f87d2b77ea3a53f68df7ebf8604a801",
    "239"
  ],
  [
    "13p1ijLwsnrcuyqcTvJXkq2ASdXqcnEBLE",
    "1ed467017f043e91ed4c44b4e8dd674db211c4e6",
    "0"
  ],
  [
    "3ALJH9Y951VCGcVZYAdpA3KchoP9McEj1G",
    "5ece0cadddc415b1980f001785947120acdb36fc",
    "5"
  ]
];

Base58CheckCodec _codec = new Base58CheckCodec.bitcoin();

main() {
  group("core.Base58Check", () {
//    test("encode", () => _testEncode());
//    test("decode", () => _testDecode());
    test("test vectors", () {
      for (List<String> vector in _bitcoinVectors) {
        String encoding = vector[0];
        List<int> bytes = hexToBytes(vector[1]);
        int version = int.parse(vector[2]);
        Base58CheckPayload payload = new Base58CheckPayload(version, bytes);
        expect(_codec.encode(payload), equals(encoding),
            reason: "encode " + vector.toString());
        expect(_codec.decode(encoding), equals(payload),
            reason: "decode " + vector.toString());
      }
    });
  });
}


lib/base58.dart
library base58check.base58;

import "dart:convert";
import "dart:typed_data";

/**
 * Encode and decode bytes to base58 strings.
 *
 * An alphabet must be provided.
 *
 * In order to comply with Bitcoin and Ripple standard encoding Base58Check,
 * use [Base58CheckCodec].
 */
class Base58Codec extends Codec<List<int>, String> {
  final String alphabet;

  const Base58Codec(String this.alphabet);

  @override
  Converter<List<int>, String> get encoder => new Base58Encoder(alphabet);

  @override
  Converter<String, List<int>> get decoder => new Base58Decoder(alphabet);
}

/**
 * Encode bytes to a base58 string.
 */
class Base58Encoder extends Converter<List<int>, String> {
  final String alphabet;

  const Base58Encoder(String this.alphabet);

  @override
  String convert(List<int> bytes) {
    if (bytes.length == 0) return "";

    // copy bytes because we are going to change it
    bytes = new Uint8List.fromList(bytes);

    // count number of leading zeros
    int leadingZeroes = 0;
    while (leadingZeroes < bytes.length && bytes[leadingZeroes] == 0)
      leadingZeroes++;

    String output = "";
    int startAt = leadingZeroes;
    while (startAt < bytes.length) {
      int mod = _divmod58(bytes, startAt);
      if (bytes[startAt] == 0) startAt++;
      output = alphabet[mod] + output;
    }

    if (output.length > 0) {
      while (output[0] == alphabet[0])
        output = output.substring(1, output.length);
    }
    while (leadingZeroes-- > 0) output = alphabet[0] + output;

    return output;
  }

  /**
   * number -> number / 58
   * returns number % 58
   */
  static int _divmod58(List<int> number, int startAt) {
    int remaining = 0;
    for (int i = startAt; i < number.length; i++) {
      int num = (0xFF & remaining) * 256 + number[i];
      number[i] = num ~/ 58;
      remaining = num % 58;
    }
    return remaining;
  }
}

/**
 * Decode base58 strings to bytes.
 */
class Base58Decoder extends Converter<String, List<int>> {
  final String alphabet;

  const Base58Decoder(String this.alphabet);

  @override
  List<int> convert(String input) {
    if (input.length == 0) return new Uint8List(0);

    // generate base 58 index list from input string
    List<int> input58 = new List(input.length);
    for (int i = 0; i < input.length; i++) {
      int charint = alphabet.indexOf(input[i]);
      if (charint < 0)
        throw new FormatException(
            "Invalid input formatting for Base58 decoding.");
      input58[i] = charint;
    }

    // count leading zeroes
    int leadingZeroes = 0;
    while (leadingZeroes < input58.length && input58[leadingZeroes] == 0)
      leadingZeroes++;

    // decode
    Uint8List output = new Uint8List(input.length);
    int j = output.length;
    int startAt = leadingZeroes;
    while (startAt < input58.length) {
      int mod = _divmod256(input58, startAt);
      if (input58[startAt] == 0) startAt++;
      output[--j] = mod;
    }

    // remove unnecessary leading zeroes
    while (j < output.length && output[j] == 0) j++;
    return output.sublist(j - leadingZeroes);
  }

  /**
   * number -> number / 256
   * returns number % 256
   */
  static int _divmod256(List<int> number58, int startAt) {
    int remaining = 0;
    for (int i = startAt; i < number58.length; i++) {
      int num = 58 * remaining + (number58[i] & 0xFF);
      number58[i] = num ~/ 256;
      remaining = num % 256;
    }
    return remaining;
  }
}


lib/base58check.dart
library base58check;

import "dart:convert";
import "dart:typed_data";

import "package:crypto/crypto.dart" show sha256;
import "package:collection/collection.dart" show ListEquality;

import "base58.dart";

class Base58CheckPayload {
  final int version;
  final List<int> payload;
  const Base58CheckPayload(int this.version, List<int> this.payload);
  @override
  bool operator ==(Object other) =>
      other is Base58CheckPayload &&
      version == other.version &&
      const ListEquality().equals(payload, other.payload);
  @override
  int get hashCode => version.hashCode ^ const ListEquality().hash(payload);
}

/**
 * A codec for Base58Check, a binary-to-string encoding used
 * in cryptocurrencies like Bitcoin and Ripple.
 *
 * The constructor requires the alphabet and a function that
 * performs a SINGLE-round SHA-256 digest on a [List<int>] and
 * returns a [List<int>] as result.
 *
 * For all details about Base58Check, see the Bitcoin wiki page:
 * https://en.bitcoin.it/wiki/Base58Check_encoding
 */
class Base58CheckCodec extends Codec<Base58CheckPayload, String> {
  static const BITCOIN_ALPHABET =
      "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

  final String alphabet;

  Base58CheckEncoder _encoder;
  Base58CheckDecoder _decoder;

  Base58CheckCodec(String this.alphabet) :
        _encoder = new Base58CheckEncoder(alphabet),
        _decoder = new Base58CheckDecoder(alphabet);


  /// A codec that works with the Bitcoin alphabet and the SHA256 hash function.
  Base58CheckCodec.bitcoin() : this(BITCOIN_ALPHABET);


  @override
  Converter<Base58CheckPayload, String> get encoder => _encoder;

  @override
  Converter<String, Base58CheckPayload> get decoder => _decoder;

  Base58CheckPayload decodeUnchecked(String encoded) =>
      _decoder.convertUnchecked(encoded);
}

class Base58CheckEncoder extends Converter<Base58CheckPayload, String> {
  final String alphabet;

  const Base58CheckEncoder(String this.alphabet);

  @override
  String convert(Base58CheckPayload payload) {
    Uint8List bytes = new Uint8List(payload.payload.length + 1 + 4);
    bytes[0] = 0xFF & payload.version;
    bytes.setRange(1, bytes.length - 4, payload.payload);
    List<int> checksum = _hash(bytes.sublist(0, bytes.length - 4));
    bytes.setRange(bytes.length - 4, bytes.length, checksum.getRange(0, 4));
    return new Base58Encoder(alphabet).convert(bytes);
  }
}

List<int> _hash(List<int> b) => sha256.convert(sha256.convert(b).bytes).bytes;

class Base58CheckDecoder extends Converter<String, Base58CheckPayload> {
  final String alphabet;

  const Base58CheckDecoder(String this.alphabet);

  @override
  Base58CheckPayload convert(String encoded) => _convert(encoded, true);

  Base58CheckPayload convertUnchecked(String encoded) =>
      _convert(encoded, false);

  Base58CheckPayload _convert(String encoded, bool validateChecksum) {
    List<int> bytes = new Base58Decoder(alphabet).convert(encoded);
    if (bytes.length < 5) {
      throw new FormatException(
          "Invalid Base58Check encoded string: must be at least size 5");
    }
    List<int> checksum = _hash(bytes.sublist(0, bytes.length - 4));
    List<int> providedChecksum = bytes.sublist(bytes.length - 4, bytes.length);
    if (validateChecksum &&
        !new ListEquality().equals(providedChecksum, checksum.sublist(0, 4))) {
      throw new FormatException("Invalid checksum in Base58Check encoding.");
    }
    return new Base58CheckPayload(bytes[0], bytes.sublist(1, bytes.length - 4));
  }
}

//TODO consider making VersionedBase58heckCodec<List<int>, String>
